canonical/istio-operators
# feat: Add upgrade handling to istio-pilot

### Summary

This enables the istio-pilot charm to upgrade Istio by up to one minor version.  This should be merged after #191 as it uses the `Istioctl` wrapper class to execute `istioctl`.  

This is designed to [this spec](https://docs.google.com/document/d/14SLwwW3-hWR5fAGyYWYPOg4ERGqI3CQlDF13MQ52FYw/edit#)

### Known bugs

`istioctl` v1.12, 1.13, and 1.14 have a bug where `istioctl upgrade` incorrectly creates a `ValidatingWebhookConfiguration` `istiod-default-validator` that looks for `istiod` in the default `istio-system` namespace, not the namespace we set (for us, typically `kubeflow`).  This results in a failing webhook immediately after upgrading **to** any of these versions.  This does not affect upgrading to istio v1.15 or 1.16, and will be fixed for 1.12-1.14 in a follow-on PR.

### Out of scope for this PR

* Deploying the other istio-pilot versions (1.12 through 1.15)

### Testing instructions

No automated upgrade tests are included yet, but manual testing is possible.  For example, you can do:
* `charmcraft pack` the istio-pilot and istio-gateway charms here
* `juju deploy` the two charms
* `juju refresh istio-pilot --path ./istio-pilot_ubuntu-20.04-amd64.charm`  # <--this refreshes the charm to **the same** version, which will work as defined in the spec

To test actual upgrades, make copies of your `./istio-pilot_ubuntu-20.04-amd64.charm` charm file and replace the `istioctl` in those copies with different versions (get different versions from [here](https://github.com/istio/istio/releases/)).  For example, I tested with:
* istio-pilot_ubuntu-20.04-amd64_istioctl_1.11.0.charm
* istio-pilot_ubuntu-20.04-amd64_istioctl_1.15.6.charm
* istio-pilot_ubuntu-20.04-amd64_istioctl_1.16.2.charm

With these different versions of the charm, you can now test things like
* deploy 1.15.6 and refresh to 1.16.2 (should work)
* deploy 1.16.2 and refresh to 1.15.2 (should fail), then refresh again back to 1.16.2 and it should work again
* deploy 1.15.6, refresh to 1.11.0 (should fail), then refresh to 1.16.2 (should work)
* deploy to 1.11.0, refresh to 1.16.2 (should fail), then **manually** use istioctl's 1.12 through 1.15 to update istio yourself to 1.15.  Once you're at 1.15, you can `juju resolve istio-pilot` your 1.16.2 deployment, and it will finish the upgrade and start working

Please try to poke many holes, read the error messages, say where things are unclear, etc!

---

Closes #188
